File: ios/Pretext.swift
import Foundation

class Pretext: HybridPretextSpec {
    public func measure(text: String, fontFamily: String, fontSize: Double) throws -> Double {
        PretextShared.shared.measure(text: text, fontFamily: fontFamily, fontSize: fontSize)
    }

    public func measureBatch(texts: [String], fontFamily: String, fontSize: Double) throws -> [Double] {
        PretextShared.shared.measureBatch(texts: texts, fontFamily: fontFamily, fontSize: fontSize)
    }

    public func prepareParagraphs(
        texts: [String],
        style: ParagraphStyle
    ) throws -> PreparedParagraphState {
        try prepareParagraphsWithStats(texts: texts, style: style).prepared
    }

    public func prepareInlineParagraphSegments(
        paragraphsPayload: String,
        style: ParagraphStyle
    ) throws -> PreparedParagraphState {
        try prepareInlineParagraphSegmentsWithStats(
            paragraphsPayload: paragraphsPayload,
            style: style
        ).prepared
    }

    public func prepareParagraphsWithStats(
        texts: [String],
        style: ParagraphStyle
    ) throws -> PreparedParagraphResult {
        PretextShared.shared.prepareParagraphsWithStats(texts: texts, style: style)
    }

    public func prepareInlineParagraphSegmentsWithStats(
        paragraphsPayload: String,
        style: ParagraphStyle
    ) throws -> PreparedParagraphResult {
        PretextShared.shared.prepareInlineParagraphsWithStats(
            paragraphs: try materializeInlineParagraphs(paragraphsPayload: paragraphsPayload),
            style: style
        )
    }

    public func layoutParagraphs(
        preparedId: Double,
        width: Double
    ) throws -> [LaidOutParagraph] {
        try PretextShared.shared.layoutParagraphs(preparedId: preparedId, width: width)
    }

    public func layoutParagraphsMetadata(
        preparedId: Double,
        width: Double
    ) throws -> [LaidOutParagraphMetrics] {
        try PretextShared.shared.layoutParagraphsMetadata(preparedId: preparedId, width: width)
    }

    public func layoutParagraphLines(
        preparedId: Double,
        width: Double
    ) throws -> [LaidOutParagraphLines] {
        try PretextShared.shared.layoutParagraphLines(preparedId: preparedId, width: width)
    }

    public func layoutParagraphsWithRequest(
        preparedId: Double,
        request: ParagraphLayoutRequest
    ) throws -> [LaidOutParagraph] {
        try PretextShared.shared.layoutParagraphs(preparedId: preparedId, request: request)
    }

    public func layoutParagraphsMetadataWithRequest(
        preparedId: Double,
        request: ParagraphLayoutRequest
    ) throws -> [LaidOutParagraphMetrics] {
        try PretextShared.shared.layoutParagraphsMetadata(preparedId: preparedId, request: request)
    }

    public func layoutParagraphLinesWithRequest(
        preparedId: Double,
        request: ParagraphLayoutRequest
    ) throws -> [LaidOutParagraphLines] {
        try PretextShared.shared.layoutParagraphLines(preparedId: preparedId, request: request)
    }

    public func layoutParagraphLinesWithDiagnostics(
        preparedId: Double,
        request: ParagraphLayoutRequest
    ) throws -> [LaidOutParagraphLinesWithDiagnostics] {
        try PretextShared.shared.layoutParagraphLinesWithDiagnostics(
            preparedId: preparedId,
            request: request
        )
    }

    public func createParagraphLineCursor(
        preparedId: Double,
        paragraphIndex: Double,
        request: ParagraphLayoutRequest
    ) throws -> ParagraphLineCursorState {
        try PretextShared.shared.createParagraphLineCursor(
            preparedId: preparedId,
            paragraphIndex: paragraphIndex,
            request: request
        )
    }

    public func nextParagraphLine(cursorId: Double) throws -> ParagraphLineCursorStep {
        PretextShared.shared.nextParagraphLine(cursorId: cursorId)
    }

    public func releaseParagraphLineCursor(cursorId: Double) throws {
        PretextShared.shared.releaseParagraphLineCursor(cursorId: cursorId)
    }

    public func releaseParagraphs(preparedId: Double) throws {
        PretextShared.shared.releaseParagraphs(preparedId: preparedId)
    }

    private func materializeInlineParagraphs(
        paragraphsPayload: String
    ) throws -> [[InlineSegment]] {
        guard !paragraphsPayload.isEmpty else {
            return []
        }

        guard let data = paragraphsPayload.data(using: .utf8) else {
            throw NSError(
                domain: "Pretext",
                code: 1,
                userInfo: [NSLocalizedDescriptionKey: "Failed to decode inline paragraph payload."]
            )
        }

        guard let root = try JSONSerialization.jsonObject(with: data) as? [[Any]] else {
            throw NSError(
                domain: "Pretext",
                code: 2,
                userInfo: [NSLocalizedDescriptionKey: "Inline paragraph payload must be a nested array."]
            )
        }

        return root.map { paragraph in
            paragraph.compactMap { segment in
                guard let segmentObject = segment as? [String: Any] else {
                    return nil
                }

                return InlineSegment(
                    text: segmentObject["text"] as? String ?? "",
                    breakBehavior: segmentObject["breakBehavior"] as? String ?? "",
                    fontFamily: segmentObject["fontFamily"] as? String,
                    fontSize: numberValue(segmentObject["fontSize"]),
                    lineHeight: numberValue(segmentObject["lineHeight"]),
                    letterSpacing: numberValue(segmentObject["letterSpacing"]),
                    locale: segmentObject["locale"] as? String,
                    fontWeight: segmentObject["fontWeight"] as? String,
                    fontStyle: segmentObject["fontStyle"] as? String
                )
            }
        }
    }

    private func numberValue(_ value: Any?) -> Double? {
        if let number = value as? NSNumber {
            return number.doubleValue
        }

        return nil
    }
}

File: android/src/main/java/com/margelo/nitro/pretext/Pretext.kt
package com.margelo.nitro.pretext

import com.facebook.proguard.annotations.DoNotStrip
import org.json.JSONArray

@DoNotStrip
class Pretext : HybridPretextSpec() {
  override fun measure(text: String, fontFamily: String, fontSize: Double): Double {
    return PretextShared.measure(text, fontFamily, fontSize)
  }

  override fun measureBatch(texts: Array<String>, fontFamily: String, fontSize: Double): DoubleArray {
    return PretextShared.measureBatch(texts, fontFamily, fontSize)
  }

  override fun prepareParagraphs(
    texts: Array<String>,
    style: ParagraphStyle,
  ): PreparedParagraphState {
    return prepareParagraphsWithStats(texts, style).prepared
  }

  override fun prepareInlineParagraphSegments(
    paragraphsPayload: String,
    style: ParagraphStyle,
  ): PreparedParagraphState {
    return prepareInlineParagraphSegmentsWithStats(
      paragraphsPayload,
      style,
    ).prepared
  }

  override fun prepareParagraphsWithStats(
    texts: Array<String>,
    style: ParagraphStyle,
  ): PreparedParagraphResult {
    return PretextShared.prepareParagraphsWithStats(texts, style)
  }

  override fun prepareInlineParagraphSegmentsWithStats(
    paragraphsPayload: String,
    style: ParagraphStyle,
  ): PreparedParagraphResult {
    return PretextShared.prepareInlineParagraphsWithStats(
      materializeInlineParagraphs(paragraphsPayload),
      style,
    )
  }

  override fun layoutParagraphs(
    preparedId: Double,
    width: Double,
  ): Array<LaidOutParagraph> {
    return PretextShared.layoutParagraphs(preparedId, width)
  }

  override fun layoutParagraphsMetadata(
    preparedId: Double,
    width: Double,
  ): Array<LaidOutParagraphMetrics> {
    return PretextShared.layoutParagraphsMetadata(preparedId, width)
  }

  override fun layoutParagraphLines(
    preparedId: Double,
    width: Double,
  ): Array<LaidOutParagraphLines> {
    return PretextShared.layoutParagraphLines(preparedId, width)
  }

  override fun layoutParagraphsWithRequest(
    preparedId: Double,
    request: ParagraphLayoutRequest,
  ): Array<LaidOutParagraph> {
    return PretextShared.layoutParagraphsWithRequest(preparedId, request)
  }

  override fun layoutParagraphsMetadataWithRequest(
    preparedId: Double,
    request: ParagraphLayoutRequest,
  ): Array<LaidOutParagraphMetrics> {
    return PretextShared.layoutParagraphsMetadataWithRequest(preparedId, request)
  }

  override fun layoutParagraphLinesWithRequest(
    preparedId: Double,
    request: ParagraphLayoutRequest,
  ): Array<LaidOutParagraphLines> {
    return PretextShared.layoutParagraphLinesWithRequest(preparedId, request)
  }

  override fun layoutParagraphLinesWithDiagnostics(
    preparedId: Double,
    request: ParagraphLayoutRequest,
  ): Array<LaidOutParagraphLinesWithDiagnostics> {
    return PretextShared.layoutParagraphLinesWithDiagnostics(preparedId, request)
  }

  override fun layoutRichParagraphLines(
    preparedId: Double,
    request: ParagraphLayoutRequest,
  ): Array<LaidOutRichParagraphLines> {
    return PretextShared.layoutRichParagraphLines(preparedId, request)
  }

  override fun createParagraphLineCursor(
    preparedId: Double,
    paragraphIndex: Double,
    request: ParagraphLayoutRequest,
  ): ParagraphLineCursorState {
    return PretextShared.createParagraphLineCursor(preparedId, paragraphIndex, request)
  }

  override fun nextParagraphLine(cursorId: Double): ParagraphLineCursorStep {
    return PretextShared.nextParagraphLine(cursorId)
  }

  override fun releaseParagraphLineCursor(cursorId: Double) {
    PretextShared.releaseParagraphLineCursor(cursorId)
  }

  override fun releaseParagraphs(preparedId: Double) {
    PretextShared.releaseParagraphs(preparedId)
  }

  private fun materializeInlineParagraphs(
    paragraphsPayload: String,
  ): Array<Array<InlineSegment>> {
    if (paragraphsPayload.isBlank()) {
      return emptyArray()
    }

    val root = JSONArray(paragraphsPayload)
    val paragraphs = ArrayList<Array<InlineSegment>>(root.length())

    for (paragraphIndex in 0 until root.length()) {
      val paragraphJson = root.getJSONArray(paragraphIndex)
      val paragraph = ArrayList<InlineSegment>(paragraphJson.length())

      for (segmentIndex in 0 until paragraphJson.length()) {
        val segmentJson = paragraphJson.getJSONObject(segmentIndex)
        paragraph.add(
          InlineSegment(
            kind = segmentJson.optString("kind").takeIf { it.isNotEmpty() },
            text = segmentJson.optString("text"),
            breakBehavior = segmentJson.optString("breakBehavior"),
            boxId = segmentJson.optString("boxId").takeIf { it.isNotEmpty() },
            width = segmentJson.optDoubleOrNull("width"),
            height = segmentJson.optDoubleOrNull("height"),
            baseline = segmentJson.optDoubleOrNull("baseline"),
            fontFamily = segmentJson.optString("fontFamily").takeIf { it.isNotEmpty() },
            fontSize = segmentJson.optDoubleOrNull("fontSize"),
            lineHeight = segmentJson.optDoubleOrNull("lineHeight"),
            letterSpacing = segmentJson.optDoubleOrNull("letterSpacing"),
            locale = segmentJson.optString("locale").takeIf { it.isNotEmpty() },
            fontWeight = segmentJson.optString("fontWeight").takeIf { it.isNotEmpty() },
            fontStyle = segmentJson.optString("fontStyle").takeIf { it.isNotEmpty() },
          ),
        )
      }

      paragraphs.add(paragraph.toTypedArray())
    }

    return paragraphs.toTypedArray()
  }

  private fun org.json.JSONObject.optDoubleOrNull(key: String): Double? {
    return if (has(key) && !isNull(key)) getDouble(key) else null
  }
}
